gh-2932 Execute tx via relayer

elgatovital opened the pull request:
Handles #2932, #2933 and #2936

Changes proposed in this pull request:
- Adjusted relayer banner
- Show payment method on review execution screen
- Payment method selection
- Fetch remaining relays
- Adjust validation

Banner | Sponsored | EOA | 
------|-------|--------
<img src="https://user-images.githubusercontent.com/17750984/227456769-4a01ef98-5100-448a-a764-5acdfd2e6f43.png" width = 200/>       | ![image](https://user-images.githubusercontent.com/246473/228310936-5fced88b-e47e-4109-8046-d9c258f38977.png) | ![image](https://user-images.githubusercontent.com/246473/228310959-8246d9c9-7e17-428c-b116-3a79684bdcbb.png)

elgatovital committed:
gh-2932 Add relay clock icon

elgatovital committed:
gh-2932 Disable owner cleanup for security v2

elgatovital committed:
gh-2932 Add payment method selection cell

elgatovital committed:
gh-2932 Add label cell

elgatovital committed:
gh-2932 Adjust review execution cell structure

elgatovital committed:
gh-2932 Add tx payment method enum

elgatovital committed:
gh-2932 Get remaining relays and show payment method

elgatovital committed:
gh-2936 Adjust background

elgatovital committed:
gh-2932 Adjust validation

elgatovital committed:
gh-2932 Fix background

elgatovital committed:
gh-2932 Refactor

elgatovital committed:
gh-2932 Add payment method selection vc

elgatovital committed:
gh-2932 Relay transaction

elgatovital committed:
gh-2932 Adjust relay info banner

elgatovital committed:
gh-2932 Adjust banner and payment method cell

elgatovital committed:
gh-2932 Rearrange payment info cells

elgatovital committed:
gh-2932 WIP: payment method selection dialog

elgatovital committed:
gh-2932 Fetch remaining relays on tx details screen

Commit:
gh-2932 Add one signature and log response

Commit:
gh-2932 Add all signatures. Log taskId

Commit:
gh-2932 Switch payment method to EOA when remaining relays exhausted

biafra23 reviewed (approved):
Yolo review!

MouazAlzahabi commented on Multisig/Features/Connect to Web/UI/Send Transaction Request/SendTransactionContentViewController.swift:
```diff
@@ -244,6 +244,11 @@ class SendTransactionCellBuilder {
         switch model {
+
+        case .none:
+            //TODO: throw
```
Can you please implement this TODO?

biafra23 replied:
Not sure what to throw here.

MouazAlzahabi commented on Multisig/UI/Assets/AssetsViewController.swift:
```diff
@@ -100,10 +100,12 @@ class AssetsViewController: ContainerViewController {
-            present(flow: relayOnboardingFlow!)
+            //TODO: open article
+            // Educational series will be shown in V2 of the relayer
+//            relayOnboardingFlow = RelayOnboardingFlow { [unowned self] _ in
```
I would remove this commented code and keep the TODO

MouazAlzahabi commented on Multisig/UI/Transaction/ExecuteTransaction/PaymentMethodCell.swift:
```diff
@@ -0,0 +1,91 @@
+        descriptionLabel.setStyle(.subheadlineSecondary)
+        remainingRelaysLabel.setStyle(.subheadlineSecondary.color(.primary))
+
+        self.backgroundView = UIView()
```
Please remove "self."

MouazAlzahabi commented on Multisig/UI/Transaction/ExecuteTransaction/PaymentMethodCell.swift:
```diff
@@ -0,0 +1,91 @@
+        let paymentMethodLabelString = NSMutableAttributedString(string: "Sponsored by ")
+        paymentMethodLabelString.append(gnosisSymbolString)
+
+        paymentMethodLabel.attributedText = paymentMethodLabelString
```
I don't think we have to create this block each time we set reminding & total

MouazAlzahabi commented on Multisig/UI/Transaction/ExecuteTransaction/LabelCell.swift:
```diff
@@ -0,0 +1,23 @@
+
+import UIKit
+
+class LabelCell: UITableViewCell {
```
I don't get why we need this new cell, seems it's not used

MouazAlzahabi commented on Multisig/UI/Transaction/ExecuteTransaction/ReviewExecutionCellBuilder.swift:
```diff
@@ -46,42 +48,98 @@ class ReviewExecutionCellBuilder: TransactionDetailCellBuilder {
         result.append(cell)
     }
 
+    func buildLabel(text: String) {
```
This is the only place LabelCell is used, but this method is not called, if so please remove it with LabelCell files

MouazAlzahabi commented on Multisig/UI/Transaction/ExecuteTransaction/SecondaryDetailDisclosureCell.swift:
```diff
@@ -16,7 +16,6 @@ class SecondaryDetailDisclosureCell: UITableViewCell {
         super.awakeFromNib()
         cellLabel.setStyle(.headline)
         self.backgroundView = UIView()
-        setBackgroundColor(.separator)
```
Are you sure we don't need this set ?

biafra23 replied:
I didn't remove this. I have no idea. How do I find out?

biafra23 replied:
Changed to backgroundPrimary. Not sure how to verify this.

MouazAlzahabi commented on Multisig/UI/Transaction/PaymentMethod/ChoosePaymentViewController.swift:
```diff
@@ -0,0 +1,81 @@
+    }
+
+    func numberOfSections(in tableView: UITableView) -> Int {
+        return 1
```
No need for return here

MouazAlzahabi commented on Multisig/UI/Transaction/PaymentMethod/ChoosePaymentViewController.swift:
```diff
@@ -0,0 +1,81 @@
+        return cell
+    }
+
+    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
```
Please either implement the TODO or remove this method

biafra23 replied:
It will be needed in the next PR.

MouazAlzahabi commented on Multisig/UI/Transaction/PaymentMethod/ChoosePaymentViewController.swift:
```diff
@@ -0,0 +1,81 @@
+    private func buildRelayerCell(tableView: UITableView) -> UITableViewCell {
+        let cell = tableView.dequeueCell(PaymentMethodCell.self)
+        cell.accessoryType = .none
+        cell.setRelaying(4, 5)
```
Why we have this constants here? shouldn't they be fetched from somewhere ?

biafra23 replied:
I replaced the 5 with a constant. Vitaly said it is not available via API, yet.

MouazAlzahabi commented on Multisig/UI/Transaction/TransactionDetailsViewController/TransactionDetailsViewController.swift:
```diff
@@ -118,8 +120,26 @@ class TransactionDetailsViewController: LoadableViewController, UITableViewDataS
         present(vc, animated: true, completion: nil)
     }
 
+    private func updateRemainingRelays() {
```
Do we still need this logic here?, as I can see we only need it in transaction execution screen, if so, please revert changes to this file

biafra23 replied:
Removed

MouazAlzahabi commented on Multisig/Cross-layer/Analytics/TrackingEvent.swift:
```diff
@@ -337,6 +337,7 @@ enum TrackingEvent: String, Trackable {
     case relayOnboarding2                           = "screen_onboarding_relay_2"
     case relayOnboarding3                           = "screen_onboarding_relay_3"
     case relayOnboarding4                           = "screen_onboarding_relay_4"
+    case relayChoosePayment                         = "screen_exec_tx_payment"
```
Do we need tacking events for execute with relayer ?

biafra23 replied:
I added the missing values from: https://docs.google.com/spreadsheets/d/1mj9iQIhpM-Pak7lQMb1dmxhQn1QdJoJyhbc730dZNzA/edit#gid=1949479240

MouazAlzahabi commented on Multisig/UI/Transaction/ExecuteTransaction/ReviewExecutionCellBuilder.swift:
```diff
@@ -46,42 +48,98 @@ class ReviewExecutionCellBuilder: TransactionDetailCellBuilder {
+        if case let .filled(relayerInfo) = model.relayerState {
+            cell.setRelaying(relayerInfo.remainingRelays, 5)
+        }
+        cell.setBackgroundColor(.separator)
```
Sorry but I don't like to use separator color for backgrounds, please define new background color and use it instead

biafra23 replied:
According to the latest figma that was the wron color anyways :-)

MouazAlzahabi commented on Multisig/UI/Transaction/ExecuteTransaction/ReviewExecutionCellBuilder.swift:
```diff
@@ -46,42 +48,98 @@ class ReviewExecutionCellBuilder: TransactionDetailCellBuilder {
+    func buildRelayerPayment(_ model: ExecutionOptionsUIModel, tableView: UITableView) -> UITableViewCell{
+        let cell = tableView.dequeueCell(PaymentMethodCell.self)
+        if case let .filled(relayerInfo) = model.relayerState {
+            cell.setRelaying(relayerInfo.remainingRelays, 5)
```
Please extract this magic value "5" to some constant somewhere, I wish if we can make it dynamic and fetch from backend together with remaining tx value

Commit:
gh-2932 Fixes:
- Autoselection of EOA if no remaininf relayer txs left
- Sign EOA executed txs
- Execute via EOA if no relayer executions left

Commit:
gh-2932 Remove labelcell

Commit:
gh-2932 Remove cruft and move stuff around

Commit:
gh-2932 Added remaining tracking events. But not emitting them, yet

Commit:
gh-2932 separator -> background & cruft removed

MouazAlzahabi commented on Multisig/UI/Transaction/ExecuteTransaction/PaymentMethodCell.swift:
```diff
@@ -0,0 +1,90 @@
+    }
+
+    func setRelaying(_ remaining: Int, _ total: Int) {
+
```
Let's just remove this empty line

MouazAlzahabi commented on Multisig/UI/Transaction/TransactionDetailsViewController/TransactionDetailsViewController.swift:
```diff
@@ -13,12 +13,14 @@ import Version
 
 class TransactionDetailsViewController: LoadableViewController, UITableViewDataSource, UITableViewDelegate {
     var clientGatewayService = App.shared.clientGatewayService
+    var relayService = App.shared.relayService
```
We don't need this line anymore

biafra23 replied:
Is there  a way to ask XCode to highlight unused code & variables?

MouazAlzahabi commented on Multisig/UI/Transaction/TransactionDetailsViewController/TransactionDetailsViewController.swift:
```diff
@@ -13,12 +13,14 @@ import Version
     private var reloadDataTask: URLSessionTask?
     private var confirmDataTask: URLSessionTask?
     private var rejectTask: URLSessionTask?
+    private var remainingRelaysTask: URLSessionTask?
```
We don't need this line anymore

MouazAlzahabi commented on Multisig/UI/Transaction/TransactionDetailsViewController/TransactionDetailsViewController.swift:
```diff
@@ -568,6 +570,7 @@ class TransactionDetailsViewController: LoadableViewController, UITableViewDataS
     }
 
     // returns true if the app has means to execute the transaction and the transaction has all required confirmations
+    //TODO: check remaining relays
```
Why we need to do so?

biafra23 replied:
If we do not have an EOA with funds but remaining relays left, the status should turn to Execution required.
I would like to postpone this to a future PR.

MouazAlzahabi commented on Multisig/UI/Transaction/ExecuteTransaction/TransactionExecutionController.swift:
```diff
@@ -624,6 +692,10 @@ class TransactionExecutionController {
     }
+
+    func didRelayTransaction(taskId: String) {
+
```
I guess this is to be implemented in the next pr right ?

biafra23 replied:
I think it was meant for logging the taskId. I removed the comment and the function

MouazAlzahabi commented on Multisig/UI/Transaction/ExecuteTransaction/TransactionExecutionController.swift:
```diff
@@ -601,6 +623,52 @@ class TransactionExecutionController {
+            case .success(let result):
+                LogService.shared.debug("---> taskId: \(try! response.get().taskId)")
+                DispatchQueue.main.async {
+                    //self.didRelayTransaction(taskId: <#T##String#>)
```
This commented line is a bit strange, can we remove it ?

MouazAlzahabi commented on Multisig/UI/Transaction/ExecuteTransaction/TransactionExecutionController.swift:
```diff
@@ -601,6 +623,52 @@ class TransactionExecutionController {
     }
 
+    func relay(completion: @escaping (Result<Void, Error>) -> Void) -> URLSessionTask? {
+
```
Extra line

MouazAlzahabi commented on Multisig/UI/Transaction/ExecuteTransaction/TransactionExecutionController.swift:
```diff
@@ -374,17 +394,19 @@ class TransactionExecutionController {
-            return
-        }
+            guard let requiredBalance = requiredBalance else {
+                return
```
Do we have to set error message same as below ?

MouazAlzahabi commented on Multisig/UI/Transaction/ExecuteTransaction/TransactionExecutionController.swift:
```diff
@@ -374,17 +394,19 @@ class TransactionExecutionController {
-        }
+        if remainingRelays <= ReviewExecutionViewController.MIN_RELAY_TXS_LEFT {  // TODO double check if <
+            guard let key = selectedKey, let keyBalance = key.balance.amount else {
+                return
```
Do we have to set error message same as below ?

biafra23 replied:
I would merge them all into one guard. With the error message as below.

MouazAlzahabi commented on Multisig/UI/Transaction/ExecuteTransaction/TransactionExecutionController.swift:
```diff
@@ -374,17 +394,19 @@ class TransactionExecutionController {
-        guard let key = selectedKey, let keyBalance = key.balance.amount else {
-            return
-        }
+        if remainingRelays <= ReviewExecutionViewController.MIN_RELAY_TXS_LEFT {  // TODO double check if <
```
Shouldn't remainingRelays be >?

biafra23 replied:
I think it is fine for now. The checks are for the case when the relays are exhausted

MouazAlzahabi commented on Multisig/UI/Transaction/ExecuteTransaction/ReviewExecutionCellBuilder.swift:
```diff
@@ -47,41 +48,93 @@ class ReviewExecutionCellBuilder: TransactionDetailCellBuilder {
+        cell.setBackgroundColor(.backgroundPrimary)
         switch model {
+        case .none:
+            //TODO: throw
```
Is this for the next pr ?

biafra23 replied:
I changed it but forgot to push.

MouazAlzahabi commented on Multisig/UI/Transaction/ExecuteTransaction/ReviewExecutionViewController.swift:
```diff
@@ -16,6 +16,9 @@ import SafariServices
 
 class ReviewExecutionViewController: ContainerViewController, PasscodeProtecting {
 
+    static let MIN_RELAY_TXS_LEFT = 0
```
Do we need this constant? is it possible to be else than 0?

biafra23 replied:
For debugging I like to set it to 5 to force the exceution via EOA. It also helps to find all places where this is checked

Commit:
gh-2932 Remove more cruft

Commit:
gh-2932 formatting

Commit:
gh-2932 Everything leads to insufficient balancex

MouazAlzahabi commented on Multisig/UI/Transaction/ExecuteTransaction/ReviewExecutionCellBuilder.swift:
```diff
@@ -47,41 +48,93 @@ class ReviewExecutionCellBuilder: TransactionDetailCellBuilder {
+        let (advancedIndex) = (0)
+        advancedParamCell.onCellTap = { [weak self] index in
+            guard let self = self else { return }
+            switch index {
```
Just for future, this switch can be replaced by if index.row == 0
